### PR TITLE
Feature: Events: Add input transformers

### PR DESCRIPTION
**Pull request by maxhoheiser**
## Motivation
Currently, EventBus does not support [Input Transformation](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-transform-target-input.html) as mentioned in this #9918 feature request and requested by another customer. 

Input transformation allows for defining a custom template for the message body, that can be populated with up to 100 key-value pairs using JSON paths to reference values from the original message. This is used to customize the text from an event, before it is passed to the consumer specified by the respective rule. 

Input transformers are defined as InputPaths (JSON paths) and a InputTemplate string
```
{
    "InputPathsMap": {"detail-type": "$.detail-type"},
    "InputTemplate": '"This event was of <detail-type> type."',
}
```

## Changes
This PR adds a validated test for input transformation and the functionality for event bridge to add basic InputTransformers

### CODE REVIEW

File: localstack/services/events/provider.py [hunk 1]
```diff
@@ -595,10 +593,24 @@ def filter_event_with_target_input_path(target: Dict, event: Dict) -> Dict:
     return event
 
 
+def filter_event_with_input_transformer(target: Dict, event: Dict) -> Dict:
```
[the remark this answers is not in the document]
Comment (maxhoheiser) [reply localstack/services/events/provider.py:596]: resolved in: c2276889e4181b07a46035ca8bb6572f78dc1dbc

File: localstack/services/events/provider.py [hunk 2]
```diff
@@ -595,10 +593,24 @@ def filter_event_with_target_input_path(target: Dict, event: Dict) -> Dict:
...
+    input_transformer = target.get("InputTransformer")
+    if not input_transformer:
+        return event
+    input_paths = input_transformer["InputPathsMap"]
+    input_template = input_transformer["InputTemplate"]
+    for key, path in input_paths.items():
+        value = extract_jsonpath(event, path)
+        input_template = input_template.replace(f"<{key}>", value)
+    templated_event = re.sub('"', "", input_template)
```
[the remark this answers is not in the document]
Comment (maxhoheiser) [reply localstack/services/events/provider.py:605]: event bridge targets expect messages of form "message text" but input template string must be specified as '"string <placeholder>"' at least using the boto3 method or else boto will throw an error that <placeholder> is none.

File: tests/aws/services/events/conftest.py [hunk 3]
```diff
@@ -114,3 +116,135 @@ def _delete_log_group():
...
+        "resources",
+        "detail",
+    )
+    for field in expected_fields:
+        assert field in event
+
+
+@pytest.fixture
+def assert_valid_event():
+    yield _assert_valid_event
```
[the remark this answers is not in the document]
Comment (maxhoheiser) [reply tests/aws/services/events/conftest.py]: resolved in: https://github.com/localstack/localstack/commit/0b5a2323bf507a98a4e491f969b5d120790673ef

File: tests/aws/services/events/conftest.py [hunk 4]
```diff
@@ -114,3 +116,135 @@ def _delete_log_group():
...
+            )
+
+        return messages
+
+    yield _put_events_with_filter_to_sqs
+
+
+def _put_entries_assert_results_sqs(
+    events_client, sqs_client, queue_url: str, entries: list[dict], should_match: bool
+):
```
[the remark this answers is not in the document]
Comment (maxhoheiser) [reply tests/aws/services/events/conftest.py:203]: resolved in: https://github.com/localstack/localstack/commit/ee2bd59a15b1ef90c206cd74bb343fa9597bb58b

File: tests/aws/services/events/conftest.py [hunk 5]
```diff
@@ -114,3 +116,135 @@ def _delete_log_group():
...
+                )
+                if message is not None:
+                    messages.extend(message)
+        finally:
+            clean_up(
+                bus_name=bus_name,
+                rule_name=rule_name,
+                target_ids=target_id,
+                queue_url=queue_url,
+            )
```
[the remark this answers is not in the document]
Comment (maxhoheiser) [reply tests/aws/services/events/conftest.py]: I moved the clean_up call to the tear-down of the fixture
resolved in: https://github.com/localstack/localstack/commit/b7faa45e9e372d2c5c5d425b08367b9dac857991

File: localstack/services/events/provider.py [hunk 6]
```diff
@@ -595,10 +593,24 @@ def filter_event_with_target_input_path(target: Dict, event: Dict) -> Dict:
     return event
 
 
+def filter_event_with_input_transformer(target: Dict, event: Dict) -> Dict:
+    input_transformer = target.get("InputTransformer")
+    if not input_transformer:
+        return event
+    input_paths = input_transformer["InputPathsMap"]
+    input_template = input_transformer["InputTemplate"]
```
[the remark this answers is not in the document]
Comment (maxhoheiser) [reply localstack/services/events/provider.py]: resolved in: https://github.com/localstack/localstack/commit/6f9304fe9920db16fd57dc4151a8d876298b89df

File: localstack/services/events/provider.py [hunk 7]
```diff
@@ -595,10 +593,33 @@ def filter_event_with_target_input_path(target: Dict, event: Dict) -> Dict:
     return event
 
 
+def process_event_with_input_transformer(target: Dict, event: Dict) -> Dict:
+    """
+    Process the event with the input transformer of the target event,
+    by replacing the message with the populated InputTemplate.
+    docs.aws.amazon.com/eventbridge/latest/userguide/eb-transform-target-input.html
```
Comment (joe4dev) [inline localstack/services/events/provider.py:600]: very helpful to link the relevant AWS docs page here 👍

File: localstack/services/events/provider.py [hunk 8]
```diff
@@ -164,9 +164,7 @@ def func(*args, **kwargs):
                     if target.get("InputPath"):
                         event = filter_event_with_target_input_path(target, event)
                     if target.get("InputTransformer"):
```
Comment (joe4dev) [inline localstack/services/events/provider.py:166]: hint: `if input_transformer := target.get("InputTransformer"):`

File: localstack/services/events/provider.py [hunk 9]
```diff
@@ -595,10 +593,33 @@ def filter_event_with_target_input_path(target: Dict, event: Dict) -> Dict:
     return event
 
 
+def process_event_with_input_transformer(target: Dict, event: Dict) -> Dict:
```
Comment (joe4dev) [inline localstack/services/events/provider.py:596]: nit: Consider adjusting the signature to facilitate re-use and testing:
`def process_event_with_input_transformer(input_transformer: Dict, event: Dict) -> Dict:`

Passing the `input_transformer` into this method to make it easier to re-use and in general follow the "Law of Demeter" (i.e., principle of least knowledge).

File: localstack/services/events/provider.py [hunk 10]
```diff
@@ -595,10 +593,33 @@ def filter_event_with_target_input_path(target: Dict, event: Dict) -> Dict:
...
+    input_transformer = target.get("InputTransformer")
+    if not input_transformer:
+        return event
+    try:
+        input_paths = input_transformer["InputPathsMap"]
+        input_template = input_transformer["InputTemplate"]
+    except KeyError as e:
+        LOG.error("%s key does not exist in input_transformer.", e)
+        raise e
+    for key, path in input_paths.items():
```
Comment (joe4dev) [inline localstack/services/events/provider.py:611]: sidenote: let's see in a follow-up how we can extract a method to reuse this functionality in Pipes 🚀 

For example in [Pipes](https://docs.aws.amazon.com/eventbridge/latest/pipes-reference/API_CreatePipe.html), the top-level parameter is called `InputTemplate`, which confusingly contains another `InputTemplate` alongside some `*Parameters` ([depending on the target](https://docs.aws.amazon.com/eventbridge/latest/pipes-reference/API_PipeTargetParameters.html#eventbridge-Type-PipeTargetParameters-InputTemplate)).

**Review by joe4dev — COMMENTED** [review]
Looking forward to re-using the transformer in EventBridge Pipes 🚀

File: tests/aws/services/events/test_events.py [hunk 11]
```diff
@@ -2082,3 +1963,37 @@ def _get_sqs_messages():
...
+                "Source": "com.mycompany.myapp",
+                "DetailType": "customerCreated",
+                "Detail": json.dumps(event_detail),
+            }
+        ]
+        entries_asserts = [(entries, True)]
+        input_transformer = {
+            "InputPathsMap": {"detail-type": "$.detail-type"},
+            "InputTemplate": '"This event was of <detail-type> type."',
+        }
```
[the remark this answers is not in the document]
Comment (maxhoheiser) [reply tests/aws/services/events/test_events.py]: 1st issue: added multiple keys to query values for replace InputTemplate in: https://github.com/localstack/localstack/commit/5151986a8fb0e23fe056cb83f79071df3ff6b66e

[the remark this answers is not in the document]
Comment (maxhoheiser) [reply tests/aws/services/events/test_events.py]: 2nd negative example: 
- defining a key in InputTemplate that is not in InputPathMap, boto does some pre-validation and throws an error in this case
```
if report.has_errors():
>               raise ParamValidationError(report=report.generate_report())
E               botocore.exceptions.ParamValidationError: Parameter validation failed:
E               Invalid type for parameter Targets[0].InputTransformer.InputPathsMap, value: ({'detail-type': '$.detail-type', 'timestamp': '$.time'},), type: <class 'tuple'>, valid types: <class 'dict'>
```

- using a key that is not present in the message results in a templated message that is just empty for the missing key (aws validated) added in: https://github.com/localstack/localstack/commit/74d9f20b49d61509bbfdff492613b8d7b730d6fb

File: localstack/services/events/provider.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dominikschubert) [reply localstack/services/events/provider.py:613]: nit: since this isn't super obvious might be good to add as a comment as well :+1:

**Review by dominikschubert — APPROVED** [review]
Looking really good now, thanks for addressing all of the comments! :rocket:

**Pull request merged**